Math/src/main/java/org/tribuo/math/neighbour/kdtree/KDTree.java

```diff
@@ -0,0 +1,467 @@
+                median -= (newPivotIndex + 1 - left);
+                left = newPivotIndex + 1;
+            }
+        }
+    }
+
+    /**
+     * Get the index of a pivot point within the specified range in the array of {@link IntAndVector}s. The operation is performed
+     * for a specific dimension of the vectors defined by the comparator.
+     *
+     * @param points The array of {@link IntAndVector}s.
+     * @param left The left-bound to use during this operation.
+     * @param right The right-bound to use during this operation.
+     * @param dimension The current dimension being processed.
+     * @return The index of the pivot.
+     */
+    private static int getPivotPointIndex(IntAndVector[] points, int left, int right, int dimension) {
+        int midIndex = (left + right) / 2;
+
+        int lowIndex = left;
+
+        if (compareByDimension(points[lowIndex], points[midIndex], dimension) >= 0) {
+            lowIndex = midIndex;
+            midIndex = left;
+        }
+
+        // Now, points[lowIndex] < points[midIndex] for the dimension being compared.
+        // Check where points[right] fits in, for the dimension being compared.
+        if (compareByDimension(points[right], points[lowIndex], dimension) <= 0) {
+            // points[right] <= points[lowIndex] < points[midIndex] for the dimension being compared
+            return lowIndex;
+        } else if (compareByDimension(points[right], points[midIndex], dimension) <= 0) {
+            // points[lowIndex] <= points[right] < points[midIndex]
+            return right;
+        }
+
+        // points[lowIndex] < points[midIndex] < points[right]
+        return midIndex;
+    }
+
+    /**
+     * Partition the array into two parts around the provided array index. Those less than the value at the provided
+     * array index will be on the left. Those greater than or equal to the value at the provided array index will fall
+     * on the left. The operation is performed for a specific dimension of the vectors defined by the comparator.
+     *
+     * @param points The array of {@link IntAndVector}s. NOTE: The order of the array will be changed.
+     * @param left The left-bound to use during this operation.
+     * @param right The right-bound to use during this operation.
+     * @param pivotIndex Index around which the partition is being made.
+     * @param dimension The current dimension being processed.
+     * @return The correct, updated index of the pivot point.
+     */
+    private static int partitionOnIndex(IntAndVector[] points, int left, int right, int pivotIndex, int dimension) {
+        IntAndVector pivot = points[pivotIndex];
+        // for now, move the pivot to the end of the array
+        swap(points, right, pivotIndex);
+
+        int store = left;
+        for (int idx = left; idx < right; idx++) {
+            if (compareByDimension(points[idx], pivot, dimension) <= 0) {
```

Comment:
Shouldn't this be a strict less than? The docs say it's less than on the left and greater than or equal on the right.